lib/token_authenticate_me/models/authenticatable.rb

```diff
@@ -44,6 +62,27 @@ def create_reset_token!
         def password_expiration_hours
           8
         end
+
+        def password=(unencrypted_password)
+          super(unencrypted_password) unless unencrypted_password.blank? && !password_required?
+        end
+
+        def current_password_correct
+          errors.add(:current_password, 'is required to change email and/or password') if current_password.blank?
+          errors.add(:current_password, 'is incorrect') unless self.class.find(id).authenticate(current_password)
```

Comment:
Aren't we already the model instance here? Can't we just do self.authenticate(current_password)?